### PR TITLE
tests: Make use of tests/libsecret

### PR DESCRIPTION
**Pull request by ormergi**
### What this PR does
Change tests that involve defining simple Secret objects to use `tests/libsecret` package

Affected test suites:
- compute/credentials.go
- storage/export.go
- virtctl/credentials.go
- vmi_configuration_test.gp
- vmi_lifecycle_test.go

**Before this PR:**
E2E tests created the secret in-line, some added the `kubevirt.io/secret` label and some dont.

**After this PR:**
E2E tests create the secret object using libsecret, all tests use the same expected configuration.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Special notes for your reviewer
The global test suite deletes cleanup that runs after each test delete secrets who has `kubevirt.io/secret` label.
It is not clear why this filtering is necessary, therefore we should revisit it and maybe remove the filtering so that all secrets in Kubevirt tests namespaces will get deleted by the global tests suite clean up.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

**Comment by fossedihelm** [comment]
/hold
Based on https://github.com/kubevirt/kubevirt/pull/11545. Wait until it is merged

**Comment by ormergi** [comment]
rebased

**Comment by ormergi** [comment]
https://github.com/kubevirt/kubevirt/pull/11545 is merged 

/unhold

**Comment by ormergi** [comment]
sig-storage passed but it seems to failed on teardonw nhttps://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11547/pull-kubevirt-e2e-k8s-1.29-sig-storage/1783101831979732992#1:build-log.txt%3A4576-4580

File: tests/virtctl/credentials.go
Comment (EdDev) [inline tests/virtctl/credentials.go:97]: I do not see a reason to use `append` here.

File: tests/virtctl/credentials.go
Comment (EdDev) [inline tests/virtctl/credentials.go:117]: Same

File: tests/vmi_configuration_test.go
Comment (EdDev) [inline tests/vmi_configuration_test.go:654]: Please reason why this has been dropped.

**Review by EdDev — CHANGES_REQUESTED** [review]

File: tests/vmi_lifecycle_test.go
Comment (ormergi) [reply tests/vmi_lifecycle_test.go:432]: The original code,  set `kubevirt.io/secret` label with some value at line 438:  kubevirt.io/secret=nonexistent.
libesecret.New create a secret with the label with no value: kubevirt.io/secret="", which is fine because the global test suite cleanup doesnt filter by the annotation value (just by its presence)

File: tests/storage/export.go
Comment (ormergi) [reply tests/storage/export.go:240]: libsecret.New implementation use Data, StringData is just for convince, when the secret is created all its content get merged to the Data field.
https://kubernetes.io/docs/reference/kubernetes-api/config-and-storage-resources/secret-v1/#Secret

There are places where the secret data map value are strings and in other places its bytes, I dont have strong opinion what is better, seem unimportent.

File: tests/storage/export.go
Comment (ormergi) [reply tests/storage/export.go:986]: same

File: tests/virtctl/credentials.go
Comment (ormergi) [reply tests/virtctl/credentials.go:97]: Thanks for the heads up, fixed.

File: tests/virtctl/credentials.go
Comment (ormergi) [reply tests/virtctl/credentials.go:117]: Fixed

File: tests/vmi_configuration_test.go
Comment (ormergi) [reply tests/vmi_configuration_test.go:654]: The secret type Opaque is the default and defined implicitly, it seemed redundant to specify it.
https://kubernetes.io/docs/concepts/configuration/secret/#opaque-secrets

File: tests/storage/export.go
Comment (ormergi) [reply tests/storage/export.go:986]: Please see https://github.com/kubevirt/kubevirt/pull/11547#discussion_r1579668750

File: tests/vmi_lifecycle_test.go
Comment (EdDev) [reply tests/vmi_lifecycle_test.go:432]: > libesecret.New create a secret with the label with no value: kubevirt.io/secret=""

I see it adds the label with a value of its name, so I still do not understand the sentence and its relevance.

File: tests/vmi_lifecycle_test.go
Comment (EdDev) [reply tests/vmi_lifecycle_test.go:441]: Sounds reasonable.

My intention in such requests is also to mention it on the commit message.
While I want to understand myself, a future me may want to understand what was done and why.

File: tests/storage/export.go
Comment (EdDev) [reply tests/storage/export.go:240]: Please reason for this in the commit message.

The API provided both options, probably to ease the user per the usage.
But now you made it harder for the user to pass the information (as a conversion is needed).
I think you need to make it easy to consume the input without making it hard to use.

You can try and use generics to make this work.

File: tests/vmi_configuration_test.go
Comment (EdDev) [reply tests/vmi_configuration_test.go:654]: Please add it to the commit.

File: tests/vmi_lifecycle_test.go
Comment (ormergi) [reply tests/vmi_lifecycle_test.go:441]: Done

File: tests/vmi_lifecycle_test.go
Comment (ormergi) [reply tests/vmi_lifecycle_test.go:432]: My bad, I removed the comment about it from the commit message.

In follow up PR we can revisit how secret are deleted and maybe remove the filtering by the mentioned label.

File: tests/storage/export.go
Comment (ormergi) [reply tests/storage/export.go:240]: > Please reason for this in the commit message.

Done

> The API provided both options, probably to ease the user per the usage.
> But now you made it harder for the user to pass the information (as a conversion is needed).
> I think you need to make it easy to consume the input without making it hard to use.

There are places that pass strings and other pass bytes, judging which is harder depends on usage.
I planned to change all tests to use libsecret first and do some optimizations as follow up, such as getting red of secret label for tests.

To address what you said I though about introducing `addDataEntry`, `addStringDataEntry` options to make it easier to use.
If you not insist I prefer to do in on a follow up PR.

> You can try and use generics to make this work.

I dont think generics can work in this case because the given argument (that has dynamic type) has to be assigned to a strict typed field.

File: tests/vmi_configuration_test.go
Comment (ormergi) [reply tests/vmi_configuration_test.go:654]: Done

File: tests/storage/export.go
Comment (EdDev) [reply tests/storage/export.go:240]: I posted https://github.com/kubevirt/kubevirt/pull/11854 .

I suggest you cherry-pick or rebase over it. Or suggest something else that does not require repeating conversions all the time.

**Review by EdDev — COMMENTED** [review]
It looks good.
I think including the generalization I sent will be a good addition to this change.
And it will be better to include it here over going again over all the callers and change them later.

**Comment by ormergi** [comment]
> It looks good. I think including the generalization I sent will be a good addition to this change. And it will be better to include it here over going again over all the callers and change them later.

I think we can do it as a follow up, lets not block this PR on it.

**Comment by ormergi** [comment]
Rebased

**Comment by ormergi** [comment]
[Changes](https://github.com/kubevirt/kubevirt/compare/a818efdbafdb69e357e9a5868716ed3077ce5c53..ea51d286fb7578245c046c099b010ab33ba742e8): Rebased on #11854

**Comment by ormergi** [comment]
[Changes](https://github.com/kubevirt/kubevirt/compare/ea51d286fb7578245c046c099b010ab33ba742e8..1d951a7a8ee81df2df121bc47457729e93295505): Change tests to use  libsecret.StringData{} and libsecret.Data{}

File: tests/storage/export.go
Comment (ormergi) [reply tests/storage/export.go:240]: Done

This PR is now rebased on #11854

**Review by ormergi — COMMENTED** [review]
@EdDev I rebased on #11854 and change the tests accordingly, PTAL

**Comment by ormergi** [comment]
Rebased

**Review by EdDev — APPROVED** [review]

**Review by enp0s3 — COMMENTED** [review]
/approve

**Pull request merged**
